pkg/controller/grafanadashboard/dashboard_controller.go

```diff
@@ -132,6 +134,57 @@ func (r *ReconcileGrafanaDashboard) Reconcile(request reconcile.Request) (reconc
 		return r.deleteDashboard(cr)
 	}
 
+	changed, hash := r.hasDashboardChanged(cr.Spec.Json, cr)
+	if changed {
+		cr.Status.LastConfig = hash
+		r.client.Update(context.TODO(), cr)
+	}
+
+	known, err := r.helper.IsKnownDashboard(cr)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	jsonTest, err := isJSON(cr.Spec.Json)
+	// Dashboard is in cm and JSON is valid
+	if known && jsonTest == true {
+		if !changed {
+			log.Info(fmt.Sprintf("dashboard '%s' reconciled but no changes", cr.Name))
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// Dashboard is in cm and JSON is invalid
+	if jsonTest == false && changed {
+		jsonError := err.Error()
+		msg := fmt.Sprintf("Invalid JSON, Error: %s", jsonError)
+		common.AppendMessage(msg, cr)
+		if known {
+			log.Info(fmt.Sprintf("invalid JSON found in reconciled dashboard '%s'", cr.Name))
+			r.deleteDashboard(cr)
+			cr.Status.Phase = common.StatusResourceUninitialized
+			r.client.Update(context.TODO(), cr)
+			return reconcile.Result{}, nil
+		}
+		log.Info(fmt.Sprintf("invalid JSON found in dashboard '%s'", cr.Name))
+		r.client.Update(context.TODO(), cr)
+		return reconcile.Result{}, nil
+	}
+
+	if jsonTest == false && !changed {
+		if !known {
+			log.Info(fmt.Sprintf("dashboard with invalid JSON '%s' reconciled but no changes", cr.Name))
+			return reconcile.Result{}, nil
+		}
+		if known {
+			r.deleteDashboard(cr)
```

Comment:
This could lead to a valid dashboard to be removed from the configmap when a change is made that breaks the json and the dashboard is reconciled a second time (e.g. by adding a label).

I'm unsure if we should remove a dashboard from the configmap, it shouldn't be possible for it to be imported in the first place when the json was invalid.

Reply:
Yeah just typing out an answer to your last comment I realise we should leave the valid dashboard in place if the json is now invalid.